File: examples/checks/odrl_dpv_risk_ranked.py
# Independent Python checks for the odrl_dpv_risk_ranked example.
from .common import run_fragment_checks

CHECKS = [
    ('4 risk rows were derived', ['Risk: portability is restricted because exporting user data is prohibited. Clause C4: Users are not permitted to export their data.']),
    ('ranked output is in descending score order', ['Rows are sorted by descending score so the highest-risk clauses are reviewed first.']),
    ('score range is 70 to 100', ['score=100 (risk:HighRisk, risk:HighSeverity) clause C1']),
    ('high=3 moderate=1 low=0 risk levels were derived', ['Risk: account/data removal is permitted without notice safeguards (no notice constraint and no duty to inform). Clause C1: Provider may remove the user account (and associated data']),
    ('5 mitigation measures were generated', ['Each triggered rule derives a risk row with a normalized score, a source clause, and one or more mitigation measures.']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/odrl_dpv_risk_ranked.go
package main

func checkODRL(ctx *Context) []Check {
	d := ctx.M()
	rows := odrlRiskRows(d)
	reported := []riskRow{}
	for _, m := range reAll(ctx.Answer, `score=(\d+) \(risk:([A-Za-z]+), risk:[A-Za-z]+\) clause (C\d)`) {
		reported = append(reported, riskRow{m[3], parseInt(m[1]), m[2], 0})
	}
	reportOK := len(reported) == len(rows)
	for i := range reported {
		if i < len(rows) {
			reportOK = reportOK && reported[i].clause == rows[i].clause && reported[i].score == rows[i].score
		}
	}
	scoresDesc := true
	for i := 1; i < len(rows); i++ {
		scoresDesc = scoresDesc && rows[i-1].score >= rows[i].score
	}
	mitig := len(reAll(ctx.Answer, `(?m)^mitigation for clause`))
	high, mod := 0, 0
	totalMit := 0
	for _, r := range rows {
		if r.level == "HighRisk" {
			high++
		}
		if r.level == "ModerateRisk" {
			mod++
		}
		totalMit += r.mitigations
	}
	change := asMap(asMap(asMap(asMap(d["Agreement"])["Policy"])["Permissions"])[":PermChangeTerms"])
	got, _ := odrlNoticeDays(change)
	need := integer(asMap(asMap(asMap(d["Consumer"])["Needs"])[":Need_ChangeOnlyWithPriorNotice"])["MinNoticeDays"])
	return []Check{{"four risk rows are derived from the policy/profile conflict scan", len(rows) == 4 && len(reported) == 4}, {"reported rows match independently computed clauses and scores", reportOK}, {"ranked output is in descending score order", scoresDesc}, {"account/data removal without notice safeguards is the highest risk", rows[0].clause == "C1" && rows[0].score == 100}, {"user-data sharing without explicit consent is scored as high risk", anyRisk(rows, "C3", 97, "HighRisk")}, {"three-day terms-change notice is below the fourteen-day consumer requirement", got == 3 && need == 14}, {"data-export prohibition creates the portability risk row", anyRisk(rows, "C4", 70, "ModerateRisk")}, {"risk level counts recompute to high=3, moderate=1, low=0", high == 3 && mod == 1}, {"five mitigation measures are generated", mitig == totalMit && totalMit == 5}}
}
